File: server/src/models/product.rs
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

/// 가격 추세
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "TEXT", rename_all = "snake_case")]
pub enum PriceTrend {
    Rising,
    Falling,
    Stable,
}

/// products 테이블
#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct Product {
    pub id: i64,
    pub shopping_mall_id: i32,
    pub category_id: Option<i32>,
    #[serde(skip_serializing)]
    pub external_product_id: String,
    #[serde(skip_serializing)]
    pub vendor_item_id: Option<String>,
    pub product_name: String,
    pub product_url: Option<String>,
    pub image_url: Option<String>,
    pub current_price: Option<i32>,
    pub lowest_price: Option<i32>,
    pub highest_price: Option<i32>,
    pub average_price: Option<i32>,
    pub unit_type: Option<String>,
    pub unit_price: Option<Decimal>,
    pub rating: Option<Decimal>,
    pub review_count: Option<i32>,
    pub is_out_of_stock: bool,
    pub price_trend: Option<PriceTrend>,
    pub days_since_lowest: Option<i32>,
    pub drop_from_average: Option<i32>,
    pub buy_timing_score: Option<i16>,
    #[serde(skip_serializing)]
    pub sales_velocity: Option<Decimal>,
    pub first_tracked_at: Option<DateTime<Utc>>,
    pub price_updated_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// shopping_malls 테이블
#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct ShoppingMall {
    pub id: i32,
    pub name: String,
    pub code: String,
    pub base_url: String,
    pub is_active: bool,
    pub created_at: DateTime<Utc>,
}

/// categories 테이블
#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct Category {
    pub id: i32,
    pub name: String,
    pub slug: String,
    pub parent_id: Option<i32>,
    pub depth: i16,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

File: app/lib/models/product.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'product.freezed.dart';
part 'product.g.dart';

/// 가격 추세 — 서버 PriceTrend enum과 1:1 대응.
enum PriceTrend {
  @JsonValue('rising')
  rising,
  @JsonValue('falling')
  falling,
  @JsonValue('stable')
  stable,
}

/// PriceTrend → API 요청에 사용하는 snake_case 문자열 변환.
extension PriceTrendX on PriceTrend {
  String get value => switch (this) {
        PriceTrend.rising => 'rising',
        PriceTrend.falling => 'falling',
        PriceTrend.stable => 'stable',
      };
}

/// 서버 Product — 상품 상세 + 검색 결과 공용
@freezed
abstract class Product with _$Product {
  const factory Product({
    required int id,
    @JsonKey(name: 'shopping_mall_id') int? shoppingMallId,
    @JsonKey(name: 'category_id') int? categoryId,
    @JsonKey(name: 'product_name') required String productName,
    @JsonKey(name: 'product_url') String? productUrl,
    @JsonKey(name: 'image_url') String? imageUrl,
    @JsonKey(name: 'current_price') int? currentPrice,
    @JsonKey(name: 'lowest_price') int? lowestPrice,
    @JsonKey(name: 'highest_price') int? highestPrice,
    @JsonKey(name: 'average_price') int? averagePrice,
    @JsonKey(name: 'price_trend') PriceTrend? priceTrend,
    @JsonKey(name: 'buy_timing_score') int? buyTimingScore,
    @JsonKey(name: 'days_since_lowest') int? daysSinceLowest,
    @JsonKey(name: 'drop_from_average') int? dropFromAverage,
    @JsonKey(name: 'is_out_of_stock') @Default(false) bool isOutOfStock,
    @JsonKey(name: 'review_count') int? reviewCount,
    @JsonKey(name: 'price_updated_at') DateTime? priceUpdatedAt,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _Product;

  factory Product.fromJson(Map<String, dynamic> json) =>
      _$ProductFromJson(json);
}

/// URL로 상품 추가 응답 — POST /products/url
@freezed
abstract class AddProductResponse with _$AddProductResponse {
  const factory AddProductResponse({
    required int id,
    @JsonKey(name: 'product_name') required String productName,
    @JsonKey(name: 'product_url') String? productUrl,
    @JsonKey(name: 'shopping_mall_id') required int shoppingMallId,
    @JsonKey(name: 'is_new') @Default(false) bool isNew,
  }) = _AddProductResponse;

  factory AddProductResponse.fromJson(Map<String, dynamic> json) =>
      _$AddProductResponseFromJson(json);
}

/// 인기 검색어
@freezed
abstract class PopularSearch with _$PopularSearch {
  const factory PopularSearch({
    required int id,
    required String keyword,
    @JsonKey(name: 'search_count') required int searchCount,
    required int rank,
    String? trend,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _PopularSearch;

  factory PopularSearch.fromJson(Map<String, dynamic> json) =>
      _$PopularSearchFromJson(json);
}
